File: src/shell.zig
const std = @import("std");
const platform = @import("platform.zig");
const io_ctx = @import("io_ctx.zig");
const paths = @import("paths.zig");
const env = @import("env.zig");

pub const source_marker = "# dot: source shell integration";
pub const bash_rc_file = ".bashrc";
pub const zsh_rc_file = ".zshrc";
pub const fish_config_dir = "fish";
pub const fish_config_file = "config.fish";
const path_marker = "# dot: add local bin to PATH";

/// Ensure the centralized integration file is sourced from the shell's RC.
/// Idempotent.
pub fn ensureSourced(shell: platform.Shell, allocator: std.mem.Allocator) !void {
    const home = env.getenv("HOME") orelse return error.NoHome;

    const rc_path = switch (shell) {
        .bash => try std.fs.path.join(allocator, &.{ home, bash_rc_file }),
        .zsh => try std.fs.path.join(allocator, &.{ home, zsh_rc_file }),
        .fish => try std.fs.path.join(allocator, &.{ home, paths.config_dir, fish_config_dir, fish_config_file }),
        .unknown => return,
    };
    defer allocator.free(rc_path);

    const integration_path = try std.fs.path.join(
        allocator,
        &.{ home, paths.local_dir, paths.bin_dir, shell.integrationFileName() },
    );
    defer allocator.free(integration_path);

    // Ensure integration file exists (open without truncating; create only if absent)
    const integ_dir = std.fs.path.dirname(integration_path) orelse return error.InvalidIntegrationPath;
    {
        const io = io_ctx.get();
        try std.Io.Dir.cwd().createDirPath(io, integ_dir);
        const integ_file = std.Io.Dir.cwd().openFile(io, integration_path, .{}) catch |e| switch (e) {
            error.FileNotFound => try std.Io.Dir.cwd().createFile(io, integration_path, .{}),
            else => return e,
        };
        integ_file.close(io);
    }

    // Check if RC already sources our file
    const io = io_ctx.get();
    const rc_content = std.Io.Dir.cwd().openFile(io, rc_path, .{}) catch |e| switch (e) {
        error.FileNotFound => {
            // Create RC and add source line
            try appendToFile(rc_path, try buildSourceLine(shell, integration_path, allocator));
            return;
        },
        else => return e,
    };
    var rc_read_buf: [4096]u8 = undefined;
    var rc_reader = rc_content.readerStreaming(io, &rc_read_buf);
    const content = try rc_reader.interface.allocRemaining(allocator, .limited(1024 * 1024));
    rc_content.close(io);
    defer allocator.free(content);

    if (std.mem.indexOf(u8, content, source_marker) != null) {
        // Already sourced — still normalize integration file to clean up any accumulated blank lines
        try normalizeIntegrationFile(integration_path, allocator);
        return;
    }

    const source_line = try buildSourceLine(shell, integration_path, allocator);
    defer allocator.free(source_line);
    try appendToFile(rc_path, source_line);

    // Also ensure PATH is set in integration file
    try ensurePathInIntegration(shell, integration_path, allocator, home);
}

/// Add (or update) a tool's shell config section in the integration file.
/// Uses # BEGIN TOOLNAME / # END TOOLNAME markers. Idempotent.
pub fn addSection(
    shell: platform.Shell,
    tool_name: []const u8,
    config: []const u8,
    allocator: std.mem.Allocator,
) !void {
    const home = env.getenv("HOME") orelse return error.NoHome;
    const integration_path = try std.fs.path.join(
        allocator,
        &.{ home, paths.local_dir, paths.bin_dir, shell.integrationFileName() },
    );
    defer allocator.free(integration_path);

    // Normalize to uppercase for markers
    const upper_name = try allocator.dupe(u8, tool_name);
    defer allocator.free(upper_name);
    for (upper_name) |*c| c.* = std.ascii.toUpper(c.*);

    const begin_marker = try std.fmt.allocPrint(allocator, "# BEGIN {s}", .{upper_name});
    defer allocator.free(begin_marker);
    const end_marker = try std.fmt.allocPrint(allocator, "# END {s}", .{upper_name});
    defer allocator.free(end_marker);

    const io = io_ctx.get();
    const file = std.Io.Dir.cwd().openFile(io, integration_path, .{}) catch |e| switch (e) {
        error.FileNotFound => {
            const new_content = try std.fmt.allocPrint(allocator, "\n{s}\n{s}\n{s}\n", .{
                begin_marker,
                config,
                end_marker,
            });
            defer allocator.free(new_content);
            try writeFile(integration_path, new_content);
            return;
        },
        else => return e,
    };
    var add_read_buf: [4096]u8 = undefined;
    var add_reader = file.readerStreaming(io, &add_read_buf);
    const existing = try add_reader.interface.allocRemaining(allocator, .limited(4 * 1024 * 1024));
    file.close(io);
    defer allocator.free(existing);

    const new_content = try rebuildWithSection(existing, begin_marker, end_marker, config, allocator);
    defer allocator.free(new_content);
    try writeFile(integration_path, new_content);
}

/// Remove a tool's section from the integration file.
pub fn removeSection(
    shell: platform.Shell,
    tool_name: []const u8,
    allocator: std.mem.Allocator,
) !void {
    const home = env.getenv("HOME") orelse return error.NoHome;
    const integration_path = try std.fs.path.join(
        allocator,
        &.{ home, paths.local_dir, paths.bin_dir, shell.integrationFileName() },
    );
    defer allocator.free(integration_path);

    const upper_name = try allocator.dupe(u8, tool_name);
    defer allocator.free(upper_name);
    for (upper_name) |*c| c.* = std.ascii.toUpper(c.*);

    const begin_marker = try std.fmt.allocPrint(allocator, "# BEGIN {s}", .{upper_name});
    defer allocator.free(begin_marker);
    const end_marker = try std.fmt.allocPrint(allocator, "# END {s}", .{upper_name});
    defer allocator.free(end_marker);

    const io = io_ctx.get();
    const file = std.Io.Dir.cwd().openFile(io, integration_path, .{}) catch return;
    var rm_read_buf: [4096]u8 = undefined;
    var rm_reader = file.readerStreaming(io, &rm_read_buf);
    const existing = try rm_reader.interface.allocRemaining(allocator, .limited(4 * 1024 * 1024));
    file.close(io);
    defer allocator.free(existing);

    const new_content = try rebuildWithoutSection(existing, begin_marker, end_marker, allocator);
    defer allocator.free(new_content);
    try writeFile(integration_path, new_content);
}

// ─── Pure text helpers (also used in tests) ───────────────────────────────────

/// Given existing file content, insert or replace the marked section with config.
/// Returns caller-owned slice.
pub fn rebuildWithSection(
    existing: []const u8,
    begin_marker: []const u8,
    end_marker: []const u8,
    config: []const u8,
    allocator: std.mem.Allocator,
) ![]u8 {
    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(allocator);

    if (std.mem.indexOf(u8, existing, begin_marker) != null) {
        // Replace existing section
        var lines = std.mem.splitScalar(u8, existing, '\n');
        var in_section = false;
        while (lines.next()) |line| {
            if (std.mem.eql(u8, line, begin_marker)) {
                try out.appendSlice(allocator, begin_marker);
                try out.append(allocator, '\n');
                try out.appendSlice(allocator, config);
                try out.append(allocator, '\n');
                in_section = true;
            } else if (std.mem.eql(u8, line, end_marker)) {
                try out.appendSlice(allocator, end_marker);
                try out.append(allocator, '\n');
                in_section = false;
            } else if (!in_section) {
                try out.appendSlice(allocator, line);
                try out.append(allocator, '\n');
            }
        }
    } else {
        // Append new section
        try out.appendSlice(allocator, existing);
        if (existing.len > 0 and existing[existing.len - 1] != '\n') {
            try out.append(allocator, '\n');
        }
        try out.appendSlice(allocator, "\n");
        try out.appendSlice(allocator, begin_marker);
        try out.append(allocator, '\n');
        try out.appendSlice(allocator, config);
        try out.append(allocator, '\n');
        try out.appendSlice(allocator, end_marker);
        try out.append(allocator, '\n');
    }

    // Normalize: collapse multiple consecutive blank lines into one, strip trailing blank lines.
    return normalizeBlankLines(allocator, out.items);
}

/// Given existing file content, return new content with the marked section removed.
/// Returns caller-owned slice.
pub fn rebuildWithoutSection(
    existing: []const u8,
    begin_marker: []const u8,
    end_marker: []const u8,
    allocator: std.mem.Allocator,
) ![]u8 {
    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(allocator);

    var lines = std.mem.splitScalar(u8, existing, '\n');
    var in_section = false;
    while (lines.next()) |line| {
        if (std.mem.eql(u8, line, begin_marker)) {
            in_section = true;
        } else if (std.mem.eql(u8, line, end_marker)) {
            in_section = false;
        } else if (!in_section) {
            try out.appendSlice(allocator, line);
            try out.append(allocator, '\n');
        }
    }

    return normalizeBlankLines(allocator, out.items);
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

/// Collapse runs of blank lines into a single blank line and strip trailing blank lines.
/// Returns a caller-owned slice.
fn normalizeBlankLines(allocator: std.mem.Allocator, content: []const u8) ![]u8 {
    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(allocator);

    var lines = std.mem.splitScalar(u8, content, '\n');
    var consecutive_blank: usize = 0;
    while (lines.next()) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r");
        if (trimmed.len == 0) {
            consecutive_blank += 1;
        } else {
            if (consecutive_blank > 0) {
                // Emit exactly one blank line between non-blank content
                try out.appendSlice(allocator, "\n");
                consecutive_blank = 0;
            }
            try out.appendSlice(allocator, line);
            try out.append(allocator, '\n');
        }
    }
    // Strip trailing blank lines — out already ends with '\n' from last non-blank line
    return out.toOwnedSlice(allocator);
}

fn buildSourceLine(shell: platform.Shell, integration_path: []const u8, allocator: std.mem.Allocator) ![]u8 {
    _ = shell;
    return std.fmt.allocPrint(allocator, "\n{s}\nsource {s}\n", .{ source_marker, integration_path });
}

fn ensurePathInIntegration(
    shell: platform.Shell,
    integration_path: []const u8,
    allocator: std.mem.Allocator,
    home: []const u8,
) !void {
    const io = io_ctx.get();
    const file = std.Io.Dir.cwd().openFile(io, integration_path, .{}) catch return;
    var path_read_buf: [4096]u8 = undefined;
    var path_reader = file.readerStreaming(io, &path_read_buf);
    const content = path_reader.interface.allocRemaining(allocator, .limited(1024 * 1024)) catch {
        file.close(io);
        return;
    };
    file.close(io);
    defer allocator.free(content);

    if (std.mem.indexOf(u8, content, path_marker) != null) return;

    const bin_dir = try std.fs.path.join(allocator, &.{ home, paths.local_dir, paths.bin_dir });
    defer allocator.free(bin_dir);

    const path_line = try shell.pathAddSyntax(bin_dir, allocator);
    defer allocator.free(path_line);

    const addition = try std.fmt.allocPrint(allocator, "\n{s}\n{s}\n", .{ path_marker, path_line });
    defer allocator.free(addition);

    try appendToFile(integration_path, addition);
}

fn normalizeIntegrationFile(path: []const u8, allocator: std.mem.Allocator) !void {
    const io = io_ctx.get();
    const file = std.Io.Dir.cwd().openFile(io, path, .{}) catch return; // not yet created — nothing to normalize
    var norm_read_buf: [4096]u8 = undefined;
    var norm_reader = file.readerStreaming(io, &norm_read_buf);
    const existing = try norm_reader.interface.allocRemaining(allocator, .limited(4 * 1024 * 1024));
    file.close(io);
    defer allocator.free(existing);
    const cleaned = try normalizeBlankLines(allocator, existing);
    defer allocator.free(cleaned);
    try writeFile(path, cleaned);
}

fn appendToFile(path: []const u8, content: []const u8) !void {
    const io = io_ctx.get();
    const file = try std.Io.Dir.cwd().createFile(io, path, .{ .truncate = false });
    defer file.close(io);
    const offset = try file.length(io);
    try file.writePositionalAll(io, content, offset);
}

fn writeFile(path: []const u8, content: []const u8) !void {
    const io = io_ctx.get();
    const file = try std.Io.Dir.cwd().createFile(io, path, .{});
    defer file.close(io);
    var write_buf: [4096]u8 = undefined;
    var file_writer = file.writerStreaming(io, &write_buf);
    try file_writer.interface.writeAll(content);
    try file_writer.interface.flush();
}

// ─── Tests ────────────────────────────────────────────────────────────────────

test "rebuildWithSection: append to empty content" {
    const result = try rebuildWithSection(
        "",
        "# BEGIN HELM",
        "# END HELM",
        "source <(helm completion bash)",
        std.testing.allocator,
    );
    defer std.testing.allocator.free(result);
    try std.testing.expect(std.mem.indexOf(u8, result, "# BEGIN HELM") != null);
    try std.testing.expect(std.mem.indexOf(u8, result, "# END HELM") != null);
    try std.testing.expect(std.mem.indexOf(u8, result, "source <(helm completion bash)") != null);
}

test "rebuildWithSection: append to existing content without section" {
    const existing = "export PATH=$PATH:/usr/local/bin\n";
    const result = try rebuildWithSection(
        existing,
        "# BEGIN HELM",
        "# END HELM",
        "source <(helm completion bash)",
        std.testing.allocator,
    );
    defer std.testing.allocator.free(result);
    // Original content preserved
    try std.testing.expect(std.mem.indexOf(u8, result, "export PATH") != null);
    // Section appended
    try std.testing.expect(std.mem.indexOf(u8, result, "# BEGIN HELM") != null);
    try std.testing.expect(std.mem.indexOf(u8, result, "source <(helm completion bash)") != null);
}

test "rebuildWithSection: replace existing section" {
    const existing =
        "export PATH=$PATH:/usr/local/bin\n" ++
        "# BEGIN HELM\n" ++
        "source <(helm completion bash)\n" ++
        "# END HELM\n" ++
        "export EDITOR=vim\n";

    const result = try rebuildWithSection(
        existing,
        "# BEGIN HELM",
        "# END HELM",
        "source <(helm completion zsh)",
        std.testing.allocator,
    );
    defer std.testing.allocator.free(result);
    // New content present
    try std.testing.expect(std.mem.indexOf(u8, result, "source <(helm completion zsh)") != null);
    // Old content removed
    try std.testing.expect(std.mem.indexOf(u8, result, "source <(helm completion bash)") == null);
    // Surrounding content preserved
    try std.testing.expect(std.mem.indexOf(u8, result, "export PATH") != null);
    try std.testing.expect(std.mem.indexOf(u8, result, "export EDITOR=vim") != null);
}

test "rebuildWithSection: idempotent on second apply" {
    const existing = "";
    const config = "source <(helm completion bash)";

    const first = try rebuildWithSection(existing, "# BEGIN HELM", "# END HELM", config, std.testing.allocator);
    defer std.testing.allocator.free(first);
    const second = try rebuildWithSection(first, "# BEGIN HELM", "# END HELM", config, std.testing.allocator);
    defer std.testing.allocator.free(second);

    // Marker appears exactly once
    var count: usize = 0;
    var it = std.mem.splitSequence(u8, second, "# BEGIN HELM");
    while (it.next()) |_| count += 1;
    try std.testing.expectEqual(@as(usize, 2), count); // split produces n+1 parts for n occurrences
}

test "rebuildWithoutSection: removes section" {
    const existing =
        "export PATH=$PATH:/usr/local/bin\n" ++
        "# BEGIN HELM\n" ++
        "source <(helm completion bash)\n" ++
        "# END HELM\n" ++
        "export EDITOR=vim\n";

    const result = try rebuildWithoutSection(existing, "# BEGIN HELM", "# END HELM", std.testing.allocator);
    defer std.testing.allocator.free(result);

    try std.testing.expect(std.mem.indexOf(u8, result, "# BEGIN HELM") == null);
    try std.testing.expect(std.mem.indexOf(u8, result, "# END HELM") == null);
    try std.testing.expect(std.mem.indexOf(u8, result, "source <(helm completion bash)") == null);
    // Surrounding content preserved
    try std.testing.expect(std.mem.indexOf(u8, result, "export PATH") != null);
    try std.testing.expect(std.mem.indexOf(u8, result, "export EDITOR=vim") != null);
}

test "rebuildWithoutSection: no-op if section absent" {
    const existing = "export PATH=$PATH:/usr/local/bin\nexport EDITOR=vim\n";
    const result = try rebuildWithoutSection(existing, "# BEGIN HELM", "# END HELM", std.testing.allocator);
    defer std.testing.allocator.free(result);
    try std.testing.expect(std.mem.indexOf(u8, result, "export PATH") != null);
    try std.testing.expect(std.mem.indexOf(u8, result, "export EDITOR=vim") != null);
}

test "rebuildWithoutSection: empty content stays empty" {
    const result = try rebuildWithoutSection("", "# BEGIN HELM", "# END HELM", std.testing.allocator);
    defer std.testing.allocator.free(result);
    try std.testing.expectEqualStrings("", result);
}

File: src/shell.rs
use std::path::Path;

use crate::error::DotError;
use crate::platform::Shell;

pub const SOURCE_MARKER: &str = "# dot: source shell integration";
const PATH_MARKER: &str = "# dot: add local bin to PATH";

/// Ensure the shell integration file is sourced from the user's RC file. Idempotent.
pub fn ensure_sourced(shell: Shell) -> Result<(), DotError> {
    if shell == Shell::Unknown {
        return Ok(());
    }
    let home = crate::paths::home_dir()?;
    let rc_path = match shell {
        Shell::Bash => home.join(".bashrc"),
        Shell::Zsh => home.join(".zshrc"),
        Shell::Fish => home.join(".config").join("fish").join("config.fish"),
        Shell::Unknown => return Ok(()),
    };

    let integration_path = crate::paths::shell_integration_file(shell)?;

    // Ensure integration file exists
    if let Some(parent) = integration_path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    if !integration_path.exists() {
        std::fs::write(&integration_path, "")?;
    }

    // Read RC content
    let rc_content = match std::fs::read_to_string(&rc_path) {
        Ok(c) => c,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            let source_line = build_source_line(&integration_path);
            append_to_file(&rc_path, &source_line)?;
            return Ok(());
        }
        Err(e) => return Err(DotError::Io(e)),
    };

    if rc_content.contains(SOURCE_MARKER) {
        normalize_integration_file(&integration_path)?;
        return Ok(());
    }

    let source_line = build_source_line(&integration_path);
    append_to_file(&rc_path, &source_line)?;
    ensure_path_in_integration(shell, &integration_path)?;
    Ok(())
}

/// Add or replace a tool's shell section in the integration file. Idempotent.
pub fn add_section(shell: Shell, tool_name: &str, config: &str) -> Result<(), DotError> {
    let integration_path = crate::paths::shell_integration_file(shell)?;
    if let Some(parent) = integration_path.parent() {
        std::fs::create_dir_all(parent)?;
    }

    let upper = tool_name.to_uppercase();
    let begin_marker = format!("# BEGIN {upper}");
    let end_marker = format!("# END {upper}");

    let existing = if integration_path.exists() {
        std::fs::read_to_string(&integration_path)?
    } else {
        String::new()
    };

    let new_content = rebuild_with_section(&existing, &begin_marker, &end_marker, config);
    write_file(&integration_path, &new_content)?;
    Ok(())
}

/// Remove a tool's section from the integration file.
pub fn remove_section(shell: Shell, tool_name: &str) -> Result<(), DotError> {
    let integration_path = crate::paths::shell_integration_file(shell)?;
    if !integration_path.exists() {
        return Ok(());
    }

    let upper = tool_name.to_uppercase();
    let begin_marker = format!("# BEGIN {upper}");
    let end_marker = format!("# END {upper}");

    let existing = std::fs::read_to_string(&integration_path)?;
    let new_content = rebuild_without_section(&existing, &begin_marker, &end_marker);
    write_file(&integration_path, &new_content)?;
    Ok(())
}

// ─── Pure text helpers ────────────────────────────────────────────────────────

/// Insert or replace the marked section. Returns the new file content.
pub fn rebuild_with_section(
    existing: &str,
    begin_marker: &str,
    end_marker: &str,
    config: &str,
) -> String {
    let mut out = String::new();

    if existing.contains(begin_marker) {
        let mut in_section = false;
        for line in existing.split('\n') {
            if line == begin_marker {
                out.push_str(begin_marker);
                out.push('\n');
                out.push_str(config);
                out.push('\n');
                in_section = true;
            } else if line == end_marker {
                out.push_str(end_marker);
                out.push('\n');
                in_section = false;
            } else if !in_section {
                out.push_str(line);
                out.push('\n');
            }
        }
    } else {
        out.push_str(existing);
        if !existing.is_empty() && !existing.ends_with('\n') {
            out.push('\n');
        }
        out.push('\n');
        out.push_str(begin_marker);
        out.push('\n');
        out.push_str(config);
        out.push('\n');
        out.push_str(end_marker);
        out.push('\n');
    }

    normalize_blank_lines(&out)
}

/// Remove the marked section and return the new file content.
pub fn rebuild_without_section(existing: &str, begin_marker: &str, end_marker: &str) -> String {
    let mut out = String::new();
    let mut in_section = false;

    for line in existing.split('\n') {
        if line == begin_marker {
            in_section = true;
        } else if line == end_marker {
            in_section = false;
        } else if !in_section {
            out.push_str(line);
            out.push('\n');
        }
    }

    normalize_blank_lines(&out)
}

// ─── Private helpers ─────────────────────────────────────────────────────────

fn normalize_blank_lines(content: &str) -> String {
    let mut out = String::new();
    let mut consecutive_blank = 0usize;

    for line in content.split('\n') {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            consecutive_blank += 1;
        } else {
            if consecutive_blank > 0 {
                out.push('\n');
                consecutive_blank = 0;
            }
            out.push_str(line);
            out.push('\n');
        }
    }
    out
}

fn build_source_line(integration_path: &Path) -> String {
    format!("\n{SOURCE_MARKER}\nsource {}\n", integration_path.display())
}

fn ensure_path_in_integration(shell: Shell, integration_path: &Path) -> Result<(), DotError> {
    let content = if integration_path.exists() {
        std::fs::read_to_string(integration_path)?
    } else {
        String::new()
    };

    if content.contains(PATH_MARKER) {
        return Ok(());
    }

    let bin_dir = crate::paths::local_bin_dir()?;
    let path_line = shell.path_add_syntax(&bin_dir.to_string_lossy());
    let addition = format!("\n{PATH_MARKER}\n{path_line}\n");
    append_to_file(integration_path, &addition)?;
    Ok(())
}

fn normalize_integration_file(path: &Path) -> Result<(), DotError> {
    if !path.exists() {
        return Ok(());
    }
    let existing = std::fs::read_to_string(path)?;
    let cleaned = normalize_blank_lines(&existing);
    write_file(path, &cleaned)
}

fn append_to_file(path: &Path, content: &str) -> Result<(), DotError> {
    use std::io::Write;
    let mut file = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)?;
    file.write_all(content.as_bytes())?;
    Ok(())
}

fn write_file(path: &Path, content: &str) -> Result<(), DotError> {
    std::fs::write(path, content)?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rebuild_with_section_append_to_empty() {
        let result = rebuild_with_section(
            "",
            "# BEGIN HELM",
            "# END HELM",
            "source <(helm completion bash)",
        );
        assert!(result.contains("# BEGIN HELM"));
        assert!(result.contains("# END HELM"));
        assert!(result.contains("source <(helm completion bash)"));
    }

    #[test]
    fn rebuild_with_section_append_to_existing_without_section() {
        let existing = "export PATH=$PATH:/usr/local/bin\n";
        let result = rebuild_with_section(
            existing,
            "# BEGIN HELM",
            "# END HELM",
            "source <(helm completion bash)",
        );
        assert!(result.contains("export PATH"));
        assert!(result.contains("# BEGIN HELM"));
        assert!(result.contains("source <(helm completion bash)"));
    }

    #[test]
    fn rebuild_with_section_replaces_existing() {
        let existing = "export PATH=$PATH:/usr/local/bin\n# BEGIN HELM\nsource <(helm completion bash)\n# END HELM\nexport EDITOR=vim\n";
        let result = rebuild_with_section(
            existing,
            "# BEGIN HELM",
            "# END HELM",
            "source <(helm completion zsh)",
        );
        assert!(result.contains("source <(helm completion zsh)"));
        assert!(!result.contains("source <(helm completion bash)"));
        assert!(result.contains("export PATH"));
        assert!(result.contains("export EDITOR=vim"));
    }

    #[test]
    fn rebuild_with_section_idempotent() {
        let config = "source <(helm completion bash)";
        let first = rebuild_with_section("", "# BEGIN HELM", "# END HELM", config);
        let second = rebuild_with_section(&first, "# BEGIN HELM", "# END HELM", config);
        // marker appears exactly once
        assert_eq!(second.matches("# BEGIN HELM").count(), 1);
    }

    #[test]
    fn rebuild_without_section_removes() {
        let existing = "export PATH=$PATH:/usr/local/bin\n# BEGIN HELM\nsource <(helm completion bash)\n# END HELM\nexport EDITOR=vim\n";
        let result = rebuild_without_section(existing, "# BEGIN HELM", "# END HELM");
        assert!(!result.contains("# BEGIN HELM"));
        assert!(!result.contains("# END HELM"));
        assert!(!result.contains("source <(helm completion bash)"));
        assert!(result.contains("export PATH"));
        assert!(result.contains("export EDITOR=vim"));
    }

    #[test]
    fn rebuild_without_section_noop_if_absent() {
        let existing = "export PATH=$PATH:/usr/local/bin\nexport EDITOR=vim\n";
        let result = rebuild_without_section(existing, "# BEGIN HELM", "# END HELM");
        assert!(result.contains("export PATH"));
        assert!(result.contains("export EDITOR=vim"));
    }

    #[test]
    fn rebuild_without_section_empty_content() {
        let result = rebuild_without_section("", "# BEGIN HELM", "# END HELM");
        assert_eq!(result, "");
    }
}
